python/src/main/java/org/apache/apex/malhar/python/base/BasePythonExecutionOperator.java

```diff
@@ -0,0 +1,591 @@
+  {
+    List<PythonRequestResponse> stragglerResponse = new ArrayList<>();
+    getApexPythonEngine().getDelayedResponseQueue().drainTo(stragglerResponse);
+    for (PythonRequestResponse aReqResponse : stragglerResponse) {
+      responseCounterForThisWindow += 1;
+      stragglersPort.emit(aReqResponse);
+    }
+  }
+
+  /***
+   * Instantiates the configured python engine. Only in-memory implementation provided for now
+   * @param context The operator context
+   * @return The python engine
+   */
+  protected ApexPythonEngine initApexPythonEngineImpl(Context.OperatorContext context)
+  {
+    JepPythonEngine jepPythonEngine = new JepPythonEngine("" + context.getId(),workerThreadPoolSize);
+    jepPythonEngine.setSleepTimeAfterInterpreterStart(sleepTimeDuringInterpreterBoot);
+    return jepPythonEngine;
+  }
+
+  /***
+   * Instantiates the partitioner. Only Thread Starvation based partitioner for now.
+   */
+  private void initPartitioner()
+  {
+    if (partitioner == null) {
+      synchronized (objectForLocking) {
+        if (partitioner == null) {
+          switch (partitionerType) {
+            default:
+            case THREAD_STARVATION_BASED:
+              partitioner = new ThreadStarvationBasedPartitioner(this);
+              break;
+          }
+        }
+      }
+    }
+  }
+
+  /***
+   * Starts the python engine and also sleeps for sometime ( configurable) to ensure that the interpreter is
+   *  completely booted up in memory. The time taken to boot the interpreter depends on the libaries that are loaded etc
+   * @param context The Operator context
+   */
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    apexPythonEngine = initApexPythonEngineImpl(context);
+    Map<PythonInterpreterConfig,Object> preInitConfigurations = getPreInitConfigurations();
+    apexPythonEngine.preInitInterpreter(preInitConfigurations);
+    apexPythonEngine.startInterpreter();
+    apexPythonEngine.postStartInterpreter();
+    initPartitioner();
+    processPostSetUpPythonInstructions(apexPythonEngine);
+  }
+
+  @Override
+  public void teardown()
+  {
+    super.teardown();
+    apexPythonEngine.stopInterpreter();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    requestCounterForThisWindow = 0;
+    responseCounterForThisWindow = 0;
+    numNullResponsesPerWindow = 0;
+    currentWindowId = windowId;
+  }
+
+  @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    if (responseCounterForThisWindow < requestCounterForThisWindow) {
+      LOG.info("Detected stragglers and configured flag/state for blocking at end of window is " +
+          blockAtEndOfWindowForStragglers);
+      if (blockAtEndOfWindowForStragglers) {
+        while (responseCounterForThisWindow < requestCounterForThisWindow) {
+          LOG.debug("Trying to emit all stragglers before the next window can be processed");
```

Comment:
move logging outside the loop? Is there a need to add a small sleep interval to avoid the busy loop? What happens when the expected responses don't arrive, I suppose the operator will be killed by the engine and then it may be good to document this here?

Reply:
Moved logging out of the loop. Also added a small sleep to avoid the busy loop. 

This loop is triggered only if the operator user decides to wait ( by means of a configuration switch names "blockAtEndOfWindowForStragglers"). If the stragglers belonging to a window do not arrive and if the above config switch is true, the operator will eventually be killed as the endWindow() call would never return.